components/service/glean/src/test/java/mozilla/components/service/glean/storages/DatetimesStorageEngineTest.kt

```diff
@@ -0,0 +1,167 @@
+            .edit()
+            .clear()
+            .apply()
+        DatetimesStorageEngine.clearAllStores()
+    }
+
+    @Test
+    fun `datetime deserializer should correctly parse datetimes`() {
+        val persistedSample = mapOf(
+            "store1#telemetry.invalid_date" to "2019-13-01T00:00:00Z",
+            "store1#telemetry.valid_date1" to "1993-02-23T09:05:23-08:00",
+            "store1#telemetry.null" to null,
+            "store1#telemetry.invalid_string" to "test"
+        )
+
+        val storageEngine = DatetimesStorageEngineImplementation()
+
+        // Create a fake application context that will be used to load our data.
+        val context = mock(Context::class.java)
+        val sharedPreferences = mock(SharedPreferences::class.java)
+        `when`(sharedPreferences.all).thenAnswer { persistedSample }
+        `when`(context.getSharedPreferences(
+            eq(storageEngine::class.java.simpleName),
+            eq(Context.MODE_PRIVATE)
+        )).thenReturn(sharedPreferences)
+
+        storageEngine.applicationContext = context
+        val snapshot = storageEngine.getSnapshot(storeName = "store1", clearStore = true)
+        assertEquals(1, snapshot!!.size)
+        assertEquals("1993-02-23T09:05:23-08:00", snapshot["telemetry.valid_date1"])
+    }
+
+    @Test
+    fun `datetime serializer should correctly serialize datetimes`() {
+        val value = Calendar.getInstance()!!
+        value.set(1993, 1, 23, 9, 5, 23)
+        value.set(Calendar.MILLISECOND, 0)
+        value.setTimeZone(TimeZone.getTimeZone("America/Los_Angeles"))
+
+        run {
+            val storageEngine = DatetimesStorageEngineImplementation()
+            storageEngine.applicationContext = ApplicationProvider.getApplicationContext()
+
+            val metric = DatetimeMetricType(
+                disabled = false,
+                category = "telemetry",
+                lifetime = Lifetime.User,
+                name = "datetime_metric",
+                sendInPings = listOf("store1")
+            )
+
+            // Record the datetime in the store, without providing optional arguments.
+            storageEngine.set(metric, value)
+
+            // Get the snapshot from "store1"
+            val snapshot = storageEngine.getSnapshotAsJSON(storeName = "store1",
+                clearStore = true)
+            // Check that this serializes to the expected JSON format.
+            assertEquals("{\"telemetry.datetime_metric\":\"1993-02-23T09:05:23-08:00\"}",
+                snapshot.toString())
+        }
+
+        // Create a new instance of storage engine to verify serialization to storage rather than
+        // to the cache
+        run {
+            val storageEngine = DatetimesStorageEngineImplementation()
+            storageEngine.applicationContext = ApplicationProvider.getApplicationContext()
+
+            // Get the snapshot from "store1"
+            val snapshot = storageEngine.getSnapshotAsJSON(storeName = "store1",
```

Comment:
Same nit as above, can we use the test API here since you have created a `DatetimeMetricType` since you aren't specifically testing `getSnapshot()`?

Reply:
Here I want to check that the JSON is exactly what we expect it to be.  It's a subtle difference, but given that we've changed the spec to say datetimes are `(value, unit)` pairs to just `value`, I think it's good to have at least one test that looks directly at the JSON.  All of these in here probably don't need to do that though.

Reply:
Makes sense to me :)